### PR TITLE
add keyvault decrypt

### PR DESCRIPTION
**Pull request by vincentserpoul**
Nothing too fancy here, but I really quickly need the decrypt function for the vault

### CODE REVIEW

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:245]: This should be `additional_authenticated_data`. See https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/keyvault/Azure.Security.KeyVault.Keys/api/Azure.Security.KeyVault.Keys.netstandard2.0.cs for names, adjusting for idiomatic naming conventions, of course. We want Azure SDKs to be first and foremost idiomatic, but consistent second and naming is one of them.

This also shouldn't be defined on `DecryptResult`. It's passed into `DecryptParameters` (as it should be called). Specifically, see the classes at https://github.com/Azure/azure-sdk-for-net/blob/fe8e623c8ef31810a4712fb12fec4ecbc019859f/sdk/keyvault/Azure.Security.KeyVault.Keys/api/Azure.Security.KeyVault.Keys.netstandard2.0.cs#L395-L422.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:246]: Even though `aad` and `iv` shouldn't be defined here, they are base64url-encoded and should be exposed as binary, i.e. `Option<Vec<u8>>`

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:256]: Please keep consistent casing. If it's idiomatic for all-uppercase characters, do so throughout. See https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/EncryptionAlgorithm.cs for how .NET did this. IIRC, Java is all uppercase since that is idiomatic for Java statics.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:360]: These should be put in `DecryptParameters`. See https://github.com/Azure/azure-sdk-for-net/blob/fe8e623c8ef31810a4712fb12fec4ecbc019859f/sdk/keyvault/Azure.Security.KeyVault.Keys/api/Azure.Security.KeyVault.Keys.netstandard2.0.cs#L358 and https://github.com/Azure/azure-sdk-for-net/blob/fe8e623c8ef31810a4712fb12fec4ecbc019859f/sdk/keyvault/Azure.Security.KeyVault.Keys/api/Azure.Security.KeyVault.Keys.netstandard2.0.cs#L395-L415.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:356]: Even in `DecryptParameters`, `version` (not `key_version`) is optional despite what the swagger says (path variables can't be optional, but they handle this correctly on the server.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:372]: These members should be `Vec<u8>` (or `Option<Vec<u8>>`) and base64url-encoded here - don't make the user do it in their own code, which is inconsistent with our other Azure SDKs.

**Review by heaths — CHANGES_REQUESTED** [review]
Thank you for the contributoin. While we do want idiomatic code (naming conventions, usage patterns, etc.) we also want names to be consistent with existing Azure SDKs, such as the .NET SDKs I referenced a few times here, though a mix of Java and Python SDKs might be easier, but the main difference you'll see is that .NET uses PascalCase while Rust uses snake_case as you have done mostly.

File: sdk/key_vault/src/key.rs
Comment (vincentserpoul) [reply sdk/key_vault/src/key.rs:245]: I was looking at https://docs.microsoft.com/en-us/rest/api/keyvault/decrypt/decrypt#keyoperationresult
and the key operation result contain these field. I guess the azure sdk for .NET is more precise

File: sdk/key_vault/src/key.rs
Comment (vincentserpoul) [reply sdk/key_vault/src/key.rs:256]: I will follow this as well. But I was looking at the SignatureAlgorithms above, which doesn't follow the .NET.

**Comment by vincentserpoul** [comment]
I tried to look into the net, java, python, js sdk to get inspiration.
I believe some more tests on my side are in order.
I think macro could also help to be less verbose, but i am not sure you want to go down that path.

File: sdk/key_vault/src/key.rs
Comment (heaths) [reply sdk/key_vault/src/key.rs]: Yeah, we did change a few property names mainly to avoid abbreviations (we have guidelines against that). Abbreviations are great for a transform protocol but not always an SDK, or so the thinking goes. The SDK does also change the shape a bit for some things you haven't touched here, too (so not an issue with this PR).

File: sdk/key_vault/src/key.rs
Comment (heaths) [reply sdk/key_vault/src/key.rs:251]: At this point we're not concerned with breaking changes - in fact, there will be a *lot* more to come - so you could make `SignatureAlgorithms` match the guidance or ignore for now and I'll clean it up later.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:300]: I think this could all be simplified to families. RSA will only ever take the ciphertext and algorithm, while AES GCM will take the ciphertext, aad, and tag, and AES CBC the ciphertext and iv. Would it be possible to simplify these? Overall, though, I appreciate the attention to detail since the parameters do take different arguments.

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:597]: Nit: you actually don't need to do this. The service backend already treats version as optional, so even consecutive "//" won't matter.

**Review by heaths — APPROVED** [review]
Great, thanks! Before I merge, though, I wanted to give you a chance to respond to one of my comments about simplifying.

File: sdk/key_vault/src/key.rs
Comment (vincentserpoul) [reply sdk/key_vault/src/key.rs]: I simplified but had to add an error in the process, in order to make sure the params can't be created with the wrong alg. Let me know if that's ok, or even if there is a better way

**Comment by vincentserpoul** [comment]
updated with latest remarks, waiting for confirmation

File: sdk/key_vault/src/key.rs
Comment (heaths) [inline sdk/key_vault/src/key.rs:339]: Because of these two - Microsoft Crypto Board has required we don't accept `iv` for input to encrypt, and you can only provide `authentication_tag` for decrypt, can you rename these to `*DecryptParameters`? The inputs for encrypt and decrypt are different and this will at least make it clear.

**Review by heaths — CHANGES_REQUESTED** [review]
This looks great! Thank you. So much cleaner. Only one other change because inputs to encrypt and decrypt are different - at least for AES GCM. I'd just name all the structs `*DecryptParameters`. A couple may be duplicated for encryption, but should the models - or at least parameters to construct them - change in the future, we don't incur any breaking changes. I'll get it merged then.

File: sdk/key_vault/src/key.rs
Comment (vincentserpoul) [reply sdk/key_vault/src/key.rs:339]: done!

**Review by heaths — APPROVED** [review]
Excellent. Thank you!

**Pull request merged**
